File: 45.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


import bisect

def count_valid_replacements(root, new_values):
    intervals = []

    def dfs(node, low, high):
        if not node:
            return
        intervals.append((low, high))
        dfs(node.left, low, node.val)
        dfs(node.right, node.val, high)

    dfs(root, float("-inf"), float("inf"))

    new_values.sort()
    total = 0

    for low, high in intervals:
        left_idx = bisect.bisect_right(new_values, low)
        right_idx = bisect.bisect_left(new_values, high)
        total += max(0, right_idx - left_idx)

    return total


# Example usage:
# BST:       5
#           / \
#          3   8
#
root = Node(5, Node(3), Node(8))
new_vals = [1, 4, 7, 10]
print(count_valid_replacements(root, new_vals))  # Output: 4

File: 45.js
class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}


function countValidReplacements(root, newValues) {
  const intervals = [];

  function dfs(node, low, high) {
    if (!node) return;
    intervals.push([low, high]);
    dfs(node.left, low, node.val);
    dfs(node.right, node.val, high);
  }

  dfs(root, -Infinity, Infinity);
  newValues.sort((a, b) => a - b);

  let total = 0;

  function lowerBound(arr, x) {
    let l = 0, r = arr.length;
    while (l < r) {
      let mid = Math.floor((l + r) / 2);
      if (arr[mid] > x) r = mid;
      else l = mid + 1;
    }
    return l;
  }

  function upperBound(arr, x) {
    let l = 0, r = arr.length;
    while (l < r) {
      let mid = Math.floor((l + r) / 2);
      if (arr[mid] >= x) r = mid;
      else l = mid + 1;
    }
    return l;
  }

  for (let [low, high] of intervals) {
    let left = lowerBound(newValues, low);
    let right = upperBound(newValues, high);
    total += Math.max(0, right - left);
  }

  return total;
}


// Example usage:
const root = new Node(5, new Node(3), new Node(8));
console.log(countValidReplacements(root, [1, 4, 7, 10])); // 4
